mesh/spacing.hpp

```diff
@@ -0,0 +1,689 @@
+    for by Newton's method. The parameter s can be scaled. This function is not
+    nested. */
+class GeometricSpacingFunction : public SpacingFunction
+{
+public:
+   GeometricSpacingFunction(int n_, bool r_, double s_, bool scale_)
+      : SpacingFunction(n_, r_, scale_), s(s_)
+   {
+      CalculateSpacing();
+   }
+
+   virtual void SetSize(int size) override
+   {
+      n = size;
+      CalculateSpacing();
+   }
+
+   virtual void ScaleParameters(double a) override
+   {
+      if (scale)
+      {
+         s *= a;
+         CalculateSpacing();
+      }
+   }
+
+   virtual double Eval(int p) override
+   {
+      const int i = reverse ? n - 1 - p : p;
+      return n == 1 ? 1.0 : s * std::pow(r, i);
+   }
+
+   virtual void Print(std::ostream &os) const override
+   {
+      // SPACING_TYPE numIntParam numDoubleParam {int params} {double params}
+      os << GEOMETRIC << " 3 1 " << n << " " << (int) reverse << " "
+         << (int) scale << " " << s << "\n";
+   }
+
+   virtual SPACING_TYPE SpacingType() const override { return GEOMETRIC; }
+   virtual int NumIntParameters() const override { return 3; }
+   virtual int NumDoubleParameters() const override { return 1; }
+
+   virtual void GetIntParameters(Array<int> & p) const override
+   {
+      p.SetSize(3);
+      p[0] = n;
+      p[1] = (int) reverse;
+      p[2] = (int) scale;
+   }
+
+   virtual void GetDoubleParameters(Vector & p) const override
+   {
+      p.SetSize(1);
+      p[0] = s;
+   }
+
+   virtual bool Nested() const override { return false; }
+
+   virtual SpacingFunction *Clone() const override
+   {
+      return new GeometricSpacingFunction(n, reverse, s, scale);
+   }
+
+private:
+   double s;  // Initial spacing
+   double r;  // Ratio
+
+   void CalculateSpacing();
+};
+
+/** @brief Bell spacing function, which produces spacing resembling a Bell
+    curve.
+
+    The widths of the first and last intervals (elements) are prescribed, and
+    the remaining interior spacings are computed by an algorithm that minimizes
+    the ratios of adjacent spacings. If the first and last intervals are wide
+    enough, the spacing may decrease in the middle of the domain. The first and
+    last interval widths can be scaled. This function is not nested.
+ */
+class BellSpacingFunction : public SpacingFunction
+{
+public:
+   /** @brief Constructor for BellSpacingFunction.
+   @param[in] n_  Size or number of intervals, which defines elements.
+   @param[in] r_  Whether to reverse the spacings.
+   @param[in] s0_ Width of the first interval (element).
+   @param[in] s1_ Width of the last interval (element).
+   @param[in] s_  Whether to scale parameters by the refinement or coarsening
+                  factor, in the function @a SpacingFunction::ScaleParameters.
+   */
+   BellSpacingFunction(int n_, bool r_, double s0_, double s1_, bool s_)
+      : SpacingFunction(n_, r_, s_), s0(s0_), s1(s1_)
```

Comment:
Trailing `_` is generally used for denoting a private member variable, so it's a bit unorthodox in an argument.

Reply:
I removed all the `_`. You are right that C++ allows for the same name in the argument as the member variable, although I usually avoid using the same names, to minimize confusion by avoiding the question of whether it should work. Similarly, I know the order of arithmetic operations but always use parentheses, even when not needed. In any case, I agree that removing the `_` looks better.